### PR TITLE
Add documentation tool

### PR DESCRIPTION
**Pull request by nobodywasishere**
Implements #8276

Example connected to development version of vscode extension

image

### CODE REVIEW

File: src/compiler/crystal/tools/documentation.cr [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    def visit(node : Call)
+      return contains_target(node) unless node.location && @target_location.between?(node.name_location, node.name_end_location)
+
+      if target_defs = node.target_defs
+        target_defs.each do |target_def|
+          if doc = target_def.doc
+            @documentations << {doc, target_def.location.not_nil!}
+          end
+        end
+      end
```
Comment (Sija) [inline src/compiler/crystal/tools/documentation.cr:61]: **nitpick**: remove one level of nesting.

```suggestion
      node.target_defs.try &.each do |target_def|
        if doc = target_def.doc
          @documentations << {doc, target_def.location.not_nil!}
        end
      end
```

File: src/compiler/crystal/tools/documentation.cr [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        column = loc.column_number
+        filename = f
+      elsif f.is_a?(VirtualFile)
+        macro_location = f.macro.location.not_nil!
+        filename = macro_location.filename.to_s
+        line = macro_location.line_number + loc.line_number
+        column = loc.column_number
+      else
+        raise "not implemented"
+      end
```
Comment (Sija) [inline src/compiler/crystal/tools/documentation.cr:95]: **nitpick**: `case` seems IMO more ergonomic.

```suggestion
      case f
      when String
        line = loc.line_number
        column = loc.column_number
        filename = f
      when VirtualFile
        macro_location = f.macro.location.not_nil!
        filename = macro_location.filename.to_s
        line = macro_location.line_number + loc.line_number
        column = loc.column_number
      else
        raise "not implemented"
      end
```

File: src/compiler/crystal/tools/documentation.cr [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    def initialize(@status, @message)
+    end
+
+    def to_text(io : IO)
+      io.puts message
+      documentations.try do |arr|
+        arr.each do |doc, loc|
+          io.puts "#{loc}\n#{doc}\n"
+        end
+      end
```
Comment (Sija) [inline src/compiler/crystal/tools/documentation.cr:25]: ditto

```suggestion
      documentations.try &.each do |doc, loc|
        io.puts "#{loc}\n#{doc}\n"
      end
```

**Comment by nobodywasishere** [comment]
I want to be able to provide documentation for macros but can't seem to figure out a way to do that currently, not as familiar with the compiler as I want to be. Something like this, which currently fails:

```crystal
  it "find documentation of macro" do
    assert_documentation %(
      # ༓foo docs
      macro foo
        def bar
        end
      end

      f‸oo
      bar
    ), check_lines: false
  end
```

**Comment by straight-shoota** [comment]
Thank you for this contribution. But I think we'll need to talk a bit more about the details of this feature though, before we can commit to a specific implementation.
We ask all major feature contributions to go through a discussion of the general design first.

#8276 is merely a thin idea with very little details. That's hardly a design spec.
And actually I don't think this PR has much in common with that idea. I understand #8276 is mainly a command line interface to query contents from the documentation dump (`index.json`). While this is a tool for IDE integration.

IMO the main functionality of this PR is identifying the target def (or path) for a given source code location. Fetching the documentation for that is just a small extra on top, but wouldn't be necessary. Given the name of the target def/path, it's easy to fetch the documentation from `index.json`.
Considering that, I believe it would be quite reasonable to achieve the same result with a combination of `crystal tool implementations` to query the target def and `crystal doc` to fetch the focumentation.
This of course requires two separate compiler instances, doing double work (and taking double time). So it's not ideal. https://github.com/crystal-lang/crystal/issues/14705 would be an improvement for that.

Thinking about `crystal tool implementations`, maybe we could add an alternative output format for documentation to this command? It seems like this would be pretty similar to the proposal in this PR except that it's not a completely new tool but enhancing an existing one.

**Comment by nobodywasishere** [comment]
I think that makes sense to me. Apologies for going through with an implementation before having more discussion. 

I'll try seeing how well `crystal tool implementations` + `crystal doc` works and report back, don't want redundant tooling if we can't help it. Otherwise a `--with-doc` flag for the implementations command should be fairly easy to add.

**Comment by straight-shoota** [comment]
No need to apologize. The purpose of having a discussion fist is to protect you from wasting effort on working out an implementation if it gets rejected.

**Pull request closed without merging**